### ISSUE TITLE
Deal with classes being available via both system classpath and plugin

### ISSUE THREAD

**Issue by johann-petrak** [GateNLP/gateplugin-Python#45]
The recent changes to the poms of the plugin and the libs we depend on have re-introduced these warnings. 

```
gate.DocumentFormat is available via both the system classpath and a plugin; the plugin (creole://uk.ac.gate.plugins;format-bdoc;1.0-SNAPSHOT/) classes will be ignored
gate.creole.metadata.CreoleResource is available via both the system classpath and a plugin; the plugin (creole://uk.ac.gate.plugins;format-bdoc;1.0-SNAPSHOT/) classes will be ignored
gate.creole.metadata.AutoInstance is available via both the system classpath and a plugin; the plugin (creole://uk.ac.gate.plugins;format-bdoc;1.0-SNAPSHOT/) classes will be ignored
gate.Resource is available via both the system classpath and a plugin; the plugin (creole://uk.ac.gate.plugins;format-bdoc;1.0-SNAPSHOT/) classes will be ignored
gate.util.InvalidOffsetException is available via both the system classpath and a plugin; the plugin (creole://uk.ac.gate.plugins;format-bdoc;1.0-SNAPSHOT/) classes will be ignored
gate.util.DocumentFormatException is available via both the system classpath and a plugin; the plugin (creole://uk.ac.gate.plugins;format-bdoc;1.0-SNAPSHOT/) classes will be ignored
gate.DocumentContent is available via both the system classpath and a plugin; the plugin (creole://uk.ac.gate.plugins;format-bdoc;1.0-SNAPSHOT/) classes will be ignored
gate.creole.ResourceInstantiationException is available via both the system classpath and a plugin; the plugin (creole://uk.ac.gate.plugins;format-bdoc;1.0-SNAPSHOT/) classes will be ignored
gate.Document is available via both the system classpath and a plugin; the plugin (creole://uk.ac.gate.plugins;format-bdoc;1.0-SNAPSHOT/) classes will be ignored
gate.corpora.RepositioningInfo is available via both the system classpath and a plugin; the plugin (creole://uk.ac.gate.plugins;format-bdoc;1.0-SNAPSHOT/) classes will be ignored
org.apache.log4j.Logger is available via both the system classpath and a plugin; the plugin (creole://uk.ac.gate.plugins;format-bdoc;1.0-SNAPSHOT/) classes will be ignored
gate.DocumentExporter is available via both the system classpath and a plugin; the plugin (creole://uk.ac.gate.plugins;format-bdoc;1.0-SNAPSHOT/) classes will be ignored
gate.FeatureMap is available via both the system classpath and a plugin; the plugin (creole://uk.ac.gate.plugins;format-bdoc;1.0-SNAPSHOT/) classes will be ignored
gate.corpora.MimeType is available via both the system classpath and a plugin; the plugin (creole://uk.ac.gate.plugins;format-bdoc;1.0-SNAPSHOT/) classes will be ignored
CREOLE plugin loaded: Format: Bdoc 1.0-SNAPSHOT
org.apache.log4j.Logger is available via both the system classpath and a plugin; the plugin (creole://uk.ac.gate.plugins;python;2.0-SNAPSHOT/) classes will be ignored
org.apache.log4j.Priority is available via both the system classpath and a plugin; the plugin (creole://uk.ac.gate.plugins;python;2.0-SNAPSHOT/) classes will be ignored
org.apache.log4j.Level is available via both the system classpath and a plugin; the plugin (creole://uk.ac.gate.plugins;python;2.0-SNAPSHOT/) classes will be ignored
CREOLE plugin loaded: Python 2.0-SNAPSHOT
Plugin Python version=2.0-SNAPSHOT commit=532832f dirty=false
Lib basicdocument version=1.0-SNAPSHOT commit=107b122 dirty=false
Lib interaction version=4.1-SNAPSHOT commit=115fdca dirty=false
gate.resources.img.svg.DocumentExporterIcon is available via both the system classpath and a plugin; the plugin (creole://uk.ac.gate.plugins;format-bdoc;1.0-SNAPSHOT/) classes will be ignored
```

**Comment by johann-petrak** [comment]
These come from Format_Bdoc:

```
gate.DocumentFormat is available via both the system classpath and a plugin; the plugin (creole://uk.ac.gate.plugins;format-bdoc;1.0-SNAPSHOT/) classes will be ignored
gate.creole.metadata.CreoleResource is available via both the system classpath and a plugin; the plugin (creole://uk.ac.gate.plugins;format-bdoc;1.0-SNAPSHOT/) classes will be ignored
gate.creole.metadata.AutoInstance is available via both the system classpath and a plugin; the plugin (creole://uk.ac.gate.plugins;format-bdoc;1.0-SNAPSHOT/) classes will be ignored
gate.Resource is available via both the system classpath and a plugin; the plugin (creole://uk.ac.gate.plugins;format-bdoc;1.0-SNAPSHOT/) classes will be ignored
gate.util.InvalidOffsetException is available via both the system classpath and a plugin; the plugin (creole://uk.ac.gate.plugins;format-bdoc;1.0-SNAPSHOT/) classes will be ignored
gate.util.DocumentFormatException is available via both the system classpath and a plugin; the plugin (creole://uk.ac.gate.plugins;format-bdoc;1.0-SNAPSHOT/) classes will be ignored
gate.DocumentContent is available via both the system classpath and a plugin; the plugin (creole://uk.ac.gate.plugins;format-bdoc;1.0-SNAPSHOT/) classes will be ignored
gate.creole.ResourceInstantiationException is available via both the system classpath and a plugin; the plugin (creole://uk.ac.gate.plugins;format-bdoc;1.0-SNAPSHOT/) classes will be ignored
gate.Document is available via both the system classpath and a plugin; the plugin (creole://uk.ac.gate.plugins;format-bdoc;1.0-SNAPSHOT/) classes will be ignored
gate.corpora.RepositioningInfo is available via both the system classpath and a plugin; the plugin (creole://uk.ac.gate.plugins;format-bdoc;1.0-SNAPSHOT/) classes will be ignored
org.apache.log4j.Logger is available via both the system classpath and a plugin; the plugin (creole://uk.ac.gate.plugins;format-bdoc;1.0-SNAPSHOT/) classes will be ignored
gate.DocumentExporter is available via both the system classpath and a plugin; the plugin (creole://uk.ac.gate.plugins;format-bdoc;1.0-SNAPSHOT/) classes will be ignored
gate.FeatureMap is available via both the system classpath and a plugin; the plugin (creole://uk.ac.gate.plugins;format-bdoc;1.0-SNAPSHOT/) classes will be ignored
gate.corpora.MimeType is available via both the system classpath and a plugin; the plugin (creole://uk.ac.gate.plugins;format-bdoc;1.0-SNAPSHOT/) classes will be ignored
```

**Comment by greenwoodma** [comment]
somewhere within the plugin you've got gate-core as a compile or runtime dependency. Plugins should never need to depend on gate-core as it's a provided dependency via the parent. If you have a lib that includes gate-core as a dependency then if it's under your control it's probably easiest just to make it as provided. If not then you'll need to exclude gate-core when using that lib within the plugin.

**Comment by greenwoodma** [comment]
to be honest I'm still a bit confused by this one. Looking at the fix it seems you have to exclude gate-core as a dependency from the interactions library, but that pom doesn't seem to depend on gate anyway. It all seems rather muddled, and will be a little fragile as changes to gate-core may then also need to be excluded here each time. Wouldn't it be better to ensure gate-core is never a dependency here by making it provided in where ever it's actually coming from? Given the way you seem to be using these pieces that would seem to make the most sense and give you the most flexibility and maintainability.

**Comment by johann-petrak** [comment]
Thanks for spotting - the gate-core exclusion from the interactions dependency was copypasta. It does no harm but should not be there in the first place. 
I already made the dependency on gate-core in basicdocument provided so I think it should be possible to remove that exclusion as well ... I think I was confused because I still got those messages until I added the exclusion, but not removing it I do not get them. 

Another thing I am not sure I understand is the dependency on log4j-core and log4j-1.2-api: that should get inherited from the plugin base, right?

**Comment by greenwoodma** [comment]
I would think they should be inherited assuming you are using the 9.0-SNAPSHOT of gate-plugin-base

**Comment by johann-petrak** [comment]
I did not getting output from the logger. But the problem seems to be that the default setup of the logger is now different from what it was previously, could that be? 

It seems the default log level is now ERROR but it was INFO previously. Also the output format is different, output lines are now prefixed with datetime methodname loglevel class "-" 

Could it be that this needs to be set up differently early on when GATE initializes?

**Comment by johann-petrak** [comment]
The original issue is fixed.
Opened #47 for the logging issues.

**Comment by johann-petrak** [comment]
OK, so after depending on format_bdoc instead of the library and after a bit of logger changes, 
I now get those warnings again. Unlike before they now do not appear when either the python or format-bdoc plugin loads, only when I actually run the PR. 

The culprit is com.fasterxml.jackson.jr:jackson-jr-objects which is needed both in format-bdoc and plugin-python. This in turn depends on com.fasterxml.jackson.core:jackson-databind so when we depend on the jr-objects at runtime, the transitive dependencies clash with what gate-core already has on the system class path. 

So we need to explicitly depend now on jr-objects  but exclude the transitive dependencies we already get from core.

**Comment by johann-petrak** [comment]
Ok the jackson-jr related warnings are fixed now.

But now that I changed the logging API i use from the log4j1 bridge to SLF4j directly I get warnings about slf4j itself. So now we need to exclude this from any dependency that depends on slf4j.

**Comment by johann-petrak** [comment]
Not sure I understand this: the warning probably appears because org.zeroturnaround:zt-exec:jar:1.11 depends on slf4j-api, but when I exclude it from there, compilation fails although I thought we should inherit the slf4j-api dependency through the plugin base.

**Comment by greenwoodma** [comment]
at a rough guess you are using 9.0-SNAPSHOT of gate-plugin-base but overrideing the gate.core.version property to allow you to load it into an older version of GATE. That might explain why dependency:tree doesn't show it and why compilation fails.

**Comment by johann-petrak** [comment]
Thanks - that must be it! 
I guess that means as long as I want the cake and eat I have to add slf4j as a compile dependency myself and exclude the transitive one.

**Comment by greenwoodma** [comment]
Maybe. What's odd is that in 8.6 as far as I know we pull in SLF4J as part of the Tika dependencies so I would have expected you to have it anyway. But I don't know which version of gate.core.version you are specifying. If pre 8.6 then you may not have it already.

**Comment by greenwoodma** [comment]
Mind you given that anything prior to 9.0-SNAPSHOT is broken now anyway, is it worth just depending directly on the latest gate-plugin-base?

**Comment by johann-petrak** [comment]
Oh - good point. That did not really sink in completely yet.

**Comment by johann-petrak** [comment]
Mostly ok in the GUI, apart from org.apache.commons.io.FileUtils when saving the python program.

**Comment by johann-petrak** [comment]
Fixed by excluding from the gatelib-interaction dependeny.

What remains now are the warnings I get on the maven output, mostly about gate-core and log4j.

This seems to happen as part of the gate-maven-plugin:9.0-SNAPSHOT:DumpCreoleToXML step.

**Comment by greenwoodma** [comment]
> This seems to happen as part of the gate-maven-plugin:9.0-SNAPSHOT:DumpCreoleToXML step.

What's the error?

**Comment by johann-petrak** [comment]
@greenwoodma  similar to "WARNING] gate.creole.PackagedController is available via both the system classpath and a plugin; the plugin (/data/jenkins/jobs/gate-plugins/jobs/gateplugin-Python/branches/master/workspace/target/classes) classes will be ignored" 
You can see all of them if you search for DumpCreoleToXML in this log: https://jenkins.gate.ac.uk/job/gate-plugins/job/gateplugin-Python/job/master/46/execution/node/30/log/

**Comment by greenwoodma** [comment]
Hmm, that is odd. Looking at the logs of building some other plugins it seems to only happen when using 9.0-SNAPSHOT of `gate-plugin-base` which suggests I've broken something. I'll have a poke around

**Comment by greenwoodma** [comment]
Turns out this was a bug in DumpCreoleToXML. It seems that we have always accidentally been loading `gate-core` as part of the plugin when generating the expanded creole.xml file. In previous versions the warning messages didn't appear as there was no Log4J logging correctly configured. Now we've switched to using SLF4J the warning messages are appearing on the Maven logging output. Fixed via https://github.com/GateNLP/gate-top/commit/8307bb7611671ef5ef99bb5839a1f8e452f06f5e
